feat(api): Upgrade User & Group APIs to Version 2

dvjsharma opened the pull request:
<!-- SPDX-FileCopyrightText: © Fossology contributors

     SPDX-License-Identifier: GPL-2.0-only
-->

<!-- Please refer to CONTRIBUTING.md (https://github.com/fossology/fossology/blob/master/CONTRIBUTING.md)
before creating the pull request to make sure you follow all the standards. -->

## Description

This PR aims to upgrade the `/users` and `/group` endpoints to v2 in accordance with the new upgrade guidelines.

### Changes

* Completed all tasks mentioned in #2709.
* Updated `openapiv2.yaml` to have all documentation changes.
* Modified models & unit tests to make them version aware.


### How to Test

* Open the `openapiv2.yaml` file in Swagger Editor and test all the updated endpoints for v2 specific responses. All function as intended.
* Execute unit tests, all pass without any issues.



CC @GMishx @shaheemazmalmmd @dushimsam

dvjsharma commented:
Few points to keep in mind regarding some upgrades: 

- `/users`, POST:  Code changes are fine, but in `openapiv2.yaml`, to change `copyright_email_author` to camelCase, I modified the `Analysis` component. However, this component is used by `ScanOptions`, which is in turn used by `UploadOptions` and the `/jobs` endpoint. So this change might break down other endpoints. Need opinion on how to handle this.

- `/users/self`, GET: Changing `default_group` to camelCase matches with an already existing key named `defaultGroup`, thus overwriting it. Therefore, I modified `default_group` to `default_group_name` and it's camelCase for v2. I also updated the documentation for both v1 and v2 APIs to reflect this change, as well as updated the unit tests to accommodate this new name.

- `/users/tokens/{type}`, GET: `token_name` and `token_expire` keys were already renamed to `name` and `expire` during the token endpoint upgrade. I have updated the v1 API docs accordingly. Also, changed `active_tokens` key to camelCase as requested.

- For the unit tests, all modal unit tests concerning users and group endpoints are now version aware and verify the data format for both v1 and v2 specifications (enhanced coverage). However, I focused solely on resolving failing controller tests without making them version aware since many of them do not specifically assess the semantics of request or response objects, but rather the functionality, which remains largely consistent between v1 and v2. Need thoughts on this.

Apart from these points, everything functions as anticipated.

GMishx commented on src/www/ui/api/documentation/openapiv2.yaml:
```diff
@@ -2482,7 +2482,7 @@ paths:
                   - type: object
                     properties:
-                      default_group:
+                      defaultGroupName:
```
Would rather recommend refactoring `defaultGroup` in `#/components/schemas/User` as a string field which holds the group name rather than group id as there are no other endpoints which uses group ids. And perhaps, move `/groups/{id}` and endpoints under it to use the group names instead.

If you agree, we can document the changes in `/groups/{id}` as part of #2709.
/cc @dushimsam

dvjsharma replied:
Agreed, I'll push the changes soon.

GMishx commented:
> Few points to keep in mind regarding some upgrades:
> 
>     * `/users`, POST:  Code changes are fine, but in `openapiv2.yaml`, to change `copyright_email_author` to camelCase, I modified the `Analysis` component. However, this component is used by `ScanOptions`, which is in turn used by `UploadOptions` and the `/jobs` endpoint. So this change might break down other endpoints. Need opinion on how to handle this.

Ultimately the changes should be done in all endpoints. Following the principle of atomic change, changes required in different endpoints can be opened as an issue and dealt in a separate PR.

>     * `/users/self`, GET: Changing `default_group` to camelCase matches with an already existing key named `defaultGroup`, thus overwriting it. Therefore, I modified `default_group` to `default_group_name` and it's camelCase for v2. I also updated the documentation for both v1 and v2 APIs to reflect this change, as well as updated the unit tests to accommodate this new name.

As mentioned earlier, users/developers should not deal with raw IDs from application. Thus, I will recommend removing the group id, user id, etc. from API and replace with names (which are also unique).

>     * `/users/tokens/{type}`, GET: `token_name` and `token_expire` keys were already renamed to `name` and `expire` during the token endpoint upgrade. I have updated the v1 API docs accordingly. Also, changed `active_tokens` key to camelCase as requested.

Thanks :+1: I will test it again.

>     * For the unit tests, all modal unit tests concerning users and group endpoints are now version aware and verify the data format for both v1 and v2 specifications (enhanced coverage). However, I focused solely on resolving failing controller tests without making them version aware since many of them do not specifically assess the semantics of request or response objects, but rather the functionality, which remains largely consistent between v1 and v2. Need thoughts on this.

That makes sense. We shall cover other aspects when we start working on test cases for REST API.

dvjsharma commented:
As suggested, I have updated the `/users` and `/groups` endpoints to use unique names instead of group IDs or user IDs. The current modification is made with the consideration that v1 API calls also need to be supported.  I have also updated all model and controller unit tests, as well as documentation to reflect these changes. I made a few decisions, please take a look and comment:
* Since the request URL needed to be changed to use names instead of IDs (data type change as well), I adjusted the regular expression to include both names and IDs rather than creating 2 separate endpoints based on version. I managed the additional differentiation logic in the corresponding controllers. Does this sound good?
* For the `user` schema, I updated `defaultGroup` to display the group name instead of the Id by adjusting the model's get array function. Another (maybe better) option is to make this change universally in the `UserDao` to return name instead ID, but will require changes in core models as well as tests. Need thoughts on this.

GMishx commented on src/www/ui/api/index.php:
```diff
@@ -246,11 +246,11 @@ function (\Slim\Routing\RouteCollectorProxy $app) {
-    $app->post('', UserController::class . ':addUser');
-    $app->delete('/{id:\\d+}', UserController::class . ':deleteUser');
     $app->get('/self', UserController::class . ':getCurrentUser');
+    $app->get('[/{pathParam:[\\w\\- \\.\\d]+}]', UserController::class . ':getUsers');
+    $app->put('[/{pathParam:[\\w\\- \\.\\d]+}]', UserController::class . ':updateUser');
+    $app->post('', UserController::class . ':addUser');
+    $app->delete('/{pathParam:[\\w\\- \\.\\d]+}', UserController::class . ':deleteUser');
```
I would rather not allow whitespace in the group/user name. Please remove the whitespace from the regex.

dvjsharma replied:
By doing this, we won't be able to access `Default User` via requests.

dvjsharma replied:
Hey @GMishx. Please guide on the above.

GMishx replied:
Humm... We can rename the user to `defaultUser` or something like that. Allowing spaces will only cause problems.

dvjsharma replied:
Done

GMishx commented on src/www/ui/api/index.php:
```diff
@@ -273,12 +273,12 @@ function (\Slim\Routing\RouteCollectorProxy $app) {
-    $app->delete('/{id:\\d+}/user/{userId:\\d+}', GroupController::class . ':deleteGroupMember');
+    $app->post('/{pathParam:[\\w\\- \\.\\d]+}/user/{userPathParam:[\\w\\- \\.\\d]+}', GroupController::class . ':addMember');
+    $app->delete('/{pathParam:[\\w\\- \\.\\d]+}', GroupController::class . ':deleteGroup');
+    $app->delete('/{pathParam:[\\w\\- \\.\\d]+}/user/{userPathParam:[\\w\\- \\.\\d]+}', GroupController::class . ':deleteGroupMember');
     $app->get('/deletable', GroupController::class . ':getDeletableGroups');
-    $app->get('/{id:\\d+}/members', GroupController::class . ':getGroupMembers');
-    $app->put('/{id:\\d+}/user/{userId:\\d+}', GroupController::class . ':changeUserPermission');
+    $app->get('/{pathParam:[\\w\\- \\.\\d]+}/members', GroupController::class . ':getGroupMembers');
+    $app->put('/{pathParam:[\\w\\- \\.\\d]+}/user/{userPathParam:[\\w\\- \\.\\d]+}', GroupController::class . ':changeUserPermission');
```
I would rather not allow whitespace in the group/user name. Please remove the whitespace from the regex.

dvjsharma replied:
Done

GMishx commented on src/www/ui/api/Controllers/UserController.php:
```diff
@@ -28,6 +28,7 @@
 use Fossology\UI\Api\Models\InfoType;
 use Fossology\UI\Api\Models\TokenRequest;
 use Psr\Http\Message\ServerRequestInterface;
+use SebastianBergmann\CodeCoverage\Driver\Xdebug;
```
Please remove unused import.

GMishx reviewed (changes requested):
Minor changes needed. Rest looks good.

GMishx commented on src/www/ui/api/index.php:
```diff
@@ -246,11 +246,11 @@ function (\Slim\Routing\RouteCollectorProxy $app) {
-    $app->post('', UserController::class . ':addUser');
-    $app->delete('/{id:\\d+}', UserController::class . ':deleteUser');
     $app->get('/self', UserController::class . ':getCurrentUser');
+    $app->get('[/{pathParam:[\\w\\-\\.]+}]', UserController::class . ':getUsers');
+    $app->put('[/{pathParam:[\\w\\-\\.]+}]', UserController::class . ':updateUser');
```
However, this regex will not work for API V1 which expects the path variable to be a digit. Also, will fail for organizations which uses email id as username (does not allow `@`).

How about using `[\w\d\-\.@_]+` as regex? It will allow every word, digit and other special characters commonly used in emails (`-.@_`). Please feel free to expand and experiment.
Also, once you've settled on the pattern, put it in a variable as it is used at multiple locations and keeping track to update pattern in future will be difficult.

dvjsharma replied:
Hm..You are right. Also I think `[\w\d\-\.@_]+` looks good enough. I was searching for some more additions but those are very uncommon and might not be useful for our usecase. Although I need thoughts on finalising this. Additions I found were adding ``!#$%&'*+-/=?^_`{|}~`` and `"(),:;<>@[\]` characters for local part of the email. Domain part is already covered by the expression you suggested. WDYT?

GMishx replied:
Lets keep it simple and not do a complete validation.

GMishx commented on src/www/ui/api/Controllers/UserController.php:
```diff
@@ -176,13 +183,14 @@ public function getCurrentUser($request, $response, $args)
-    $id = intval($args['id']);
+    $apiVersion = ApiVersion::getVersion($request);
+    $id = $apiVersion == ApiVersion::V2 ? intval($this->restHelper->getUserDao()->getUserByName($args['pathParam'])['user_pk']) : intval($args['pathParam']);
     if ($this->dbHelper->doesIdExist("users","user_pk", $id)) {
```
Oops, think we never tested it. This condition needs to be inverted (user not exists) to throw the exception. Can you please do that?

```suggestion
    if (!$this->dbHelper->doesIdExist("users","user_pk", $id)) {
```

dvjsharma replied:
Ah, I believe I overlooked that detail during testing as well. Fixed

GMishx commented on src/www/ui/api/index.php:
```diff
@@ -245,12 +248,12 @@ function (\Slim\Routing\RouteCollectorProxy $app) {
+  function (\Slim\Routing\RouteCollectorProxy $app) use ($pattern) {
     $app->get('/self', UserController::class . ':getCurrentUser');
+    $app->get("[/{pathParam:$pattern]", UserController::class . ':getUsers');
+    $app->put("[/{pathParam:$pattern]", UserController::class . ':updateUser');
```
And if you are working on the previous change, please remove the `[]` from this path as the user information is a must for this endpoint unlike the `:getUsers`.

```suggestion
    $app->put("/{pathParam:$pattern", UserController::class . ':updateUser');
```

GMishx reviewed (changes requested):
@dvjsharma final 2 changes needed in the PR. Then I can merge it.
